atintegrators/BeamLoadingCavityPass.c

```diff
@@ -140,14 +146,40 @@ void BeamLoadingCavityPass(double *r_in,int num_particles,int nbunch,
                 r6[4] += kz[islice]; 
             }
         }
-        if(cavitymode==1){
-            update_vgen(vbeamk,vcavk,vgenk,phasegain,voltgain,detune_angle); 
-        }
+        
+        // First write the values to the buffer
         if(buffersize>0){
             write_buffer(vbeamk, vbeam_buffer, 2, buffersize);
             write_buffer(vgenk, vgen_buffer, 2, buffersize);
             write_buffer(vbunch, vbunch_buffer, 2*nbunch, buffersize);
+        }   
+
+        // If FBMode is set to ONETURN, then set the vbeam and move on
+        if(fbmode==1){
+            vbeam_set[0] = vbeamk[0];
+            vbeam_set[1] = vbeamk[1];        
+        }
+        // If FBMode is set to WINDOW, compute the vbeam_set from the buffer
+        
+        else if(fbmode==2){
+            // Compute the length of the buffer as we will not act until 
+            // the buffer is full. (2 arrays of vbeam and psi)
+            
+            bufferlengthnow = check_buffer_length(vbeam_buffer, buffersize, 2);
```

Comment:
I generally like to put as little as possible in the passmethod, so you may combine these 2 into a single function update vbeam_set, but this is a question of taste like this it is fine as well